galsim/bandpass.py

```diff
@@ -83,90 +83,122 @@ class Bandpass(object):
-            if self.blue_limit < (tp.x_min/wave_factor):
+            if self.blue_limit < (self._tp.x_min/self.wave_factor):
                 raise ValueError("Cannot set blue_limit to be less than throughput "
                                  + "LookupTable.x_min")
-            if self.red_limit > (tp.x_max/wave_factor):
+            if self.red_limit > (self._tp.x_max/self.wave_factor):
                 raise ValueError("Cannot set red_limit to be greater than throughput "
                                  + "LookupTable.x_max")
+            # Remove any values that are outside the limits
+            self.wave_list = self.wave_list[np.logical_and(self.wave_list >= self.blue_limit,
+                                                           self.wave_list <= self.red_limit) ]
             # Make sure that blue_limit and red_limit are part of wave_list.
+            if self.red_limit not in self.wave_list:
+                np.append(self.wave_list, self.red_limit)
             if self.blue_limit not in self.wave_list:
                 np.insert(self.wave_list, 0, self.blue_limit)
-            if self.red_limit not in self.wave_list:
-                np.insert(self.wave_list, -1, self.red_limit)
         else:
             self.wave_list = np.array([], dtype=np.float)
 
-        # Manual override!  Be careful!
-        if _wave_list is not None:
-            self.wave_list = _wave_list
 
-        self.func = lambda w: tp(np.array(w) * wave_factor)
+    def _initialize_tp(self):
+        # Turrn the input tp into a real function self.func.
+        # The function cannot be pickled, so will need to do this in setstate as well as init.
 
-        self.zeropoint = None
+        if self._tp is not None:
+            pass
+        elif isinstance(self._orig_tp, basestring):
+            import os
+            if os.path.isfile(self._orig_tp):
+                self._tp = galsim.LookupTable(file=self._orig_tp, interpolant='linear')
+            else:
+                # Evaluate the function somewhere to make sure it is valid before continuing on.
+                if self.red_limit is not None:
+                    test_wave = self.red_limit * self.wave_factor
+                elif blue_limit is not None:
+                    test_wave = self.blue_limit * self.wave_factor
+                else:
+                    # If neither `blue_limit` nor `red_limit` is defined, then the Bandpass should
+                    # be able to be evaluated at any wavelength, so check.
+                    test_wave = 700
+                try:
+                    self._tp = eval('lambda wave : ' + self._orig_tp)
+                    self._tp(test_wave)
+                except:
+                    raise ValueError(
+                        "String throughput must either be a valid filename or something that "+
+                        "can eval to a function of wave. Input provided: {0}".format(throughput))
+        else:
+            self._tp = self._orig_tp
```

Comment:
Should we not do some sanity checks of this, just like when there is string input?